File: php-api/src/Controllers/VendorController.php
<?php

declare(strict_types=1);

namespace Enkl\Api\Controllers;

use Enkl\Api\Db\Database;
use Enkl\Api\Services\VendorService;
use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;

/** Ported from Controllers/VendorController.cs. Org-Admin-only "Manage Vendors" surface — CRUD plus
 * per-Vendor API key generate/revoke, same callerOrgId() idiom as OrganisationApiKeyController. */
final class VendorController extends BaseController
{
    private function service(): VendorService
    {
        return new VendorService(Database::connection());
    }

    public function list(Request $request, Response $response): Response
    {
        return $this->json($response, $this->service()->list($this->callerOrgId($request)));
    }

    public function get(Request $request, Response $response, array $args): Response
    {
        $result = $this->service()->get($this->callerOrgId($request), $args['vendorId']);
        return $result === null ? $this->notFound($response) : $this->json($response, $result);
    }

    public function create(Request $request, Response $response): Response
    {
        return $this->json($response, $this->service()->create($this->callerOrgId($request), $this->body($request)));
    }

    public function update(Request $request, Response $response, array $args): Response
    {
        $result = $this->service()->update($this->callerOrgId($request), $args['vendorId'], $this->body($request));
        return $result === null ? $this->notFound($response) : $this->json($response, $result);
    }

    public function delete(Request $request, Response $response, array $args): Response
    {
        $deleted = $this->service()->delete($this->callerOrgId($request), $args['vendorId']);
        return $deleted ? $this->noContent($response) : $this->notFound($response);
    }

    public function generateApiKey(Request $request, Response $response, array $args): Response
    {
        $result = $this->service()->generateApiKey($this->callerOrgId($request), $args['vendorId']);
        return $result === null ? $this->notFound($response) : $this->json($response, $result);
    }

    public function revokeApiKey(Request $request, Response $response, array $args): Response
    {
        $result = $this->service()->revokeApiKey($this->callerOrgId($request), $args['vendorId']);
        return $result === null ? $this->notFound($response) : $this->json($response, $result);
    }
}

File: api/Enkl.Api/Controllers/VendorController.cs
using Enkl.Api.Auth;
using Enkl.Api.Dtos;
using Enkl.Api.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Enkl.Api.Controllers;

/// <summary>
/// Org-Admin-only "Manage Vendors" surface — CRUD plus per-Vendor API key generate/revoke, nested
/// under api/organisations/me same as OrganisationApiKeyController/OrganisationAnnouncementsController.
/// Every id re-derived against User.OrgId(), never client-supplied.
/// </summary>
[ApiController]
[Authorize(Policy = "OrgAdmin")]
[Route("api/organisations/me/vendors")]
public class VendorController : ControllerBase
{
    private readonly VendorService _vendors;

    public VendorController(VendorService vendors)
    {
        _vendors = vendors;
    }

    [HttpGet]
    public async Task<IActionResult> List()
    {
        return Ok(await _vendors.ListAsync(User.OrgId()));
    }

    [HttpGet("{vendorId:guid}")]
    public async Task<IActionResult> Get(Guid vendorId)
    {
        var result = await _vendors.GetAsync(User.OrgId(), vendorId);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpPost]
    public async Task<IActionResult> Create(CreateVendorRequest request)
    {
        return Ok(await _vendors.CreateAsync(User.OrgId(), request));
    }

    [HttpPut("{vendorId:guid}")]
    public async Task<IActionResult> Update(Guid vendorId, UpdateVendorRequest request)
    {
        var result = await _vendors.UpdateAsync(User.OrgId(), vendorId, request);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpDelete("{vendorId:guid}")]
    public async Task<IActionResult> Delete(Guid vendorId)
    {
        var deleted = await _vendors.DeleteAsync(User.OrgId(), vendorId);
        return deleted ? NoContent() : NotFound();
    }

    [HttpPost("{vendorId:guid}/api-key")]
    public async Task<IActionResult> GenerateApiKey(Guid vendorId)
    {
        var result = await _vendors.GenerateApiKeyAsync(User.OrgId(), vendorId);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpDelete("{vendorId:guid}/api-key")]
    public async Task<IActionResult> RevokeApiKey(Guid vendorId)
    {
        var result = await _vendors.RevokeApiKeyAsync(User.OrgId(), vendorId);
        return result is null ? NotFound() : Ok(result);
    }
}
